osu.Game/Database/BeatmapDatabase.cs

```diff
@@ -118,12 +118,49 @@ public void Reset()
 
         public void Import(IEnumerable<string> paths)
         {
+            Stack<BeatmapSetInfo> sets = new Stack<BeatmapSetInfo>();
+
             foreach (string p in paths)
-                Import(p);
+                try
+                {
+                    BeatmapSetInfo set = getBeatmapSet(p);
+
+                    sets.Push(set);
+
+                    // We may or may not want to delete the file depending on where it is stored.
+                    //  e.g. reconstructing/repairing database with beatmaps from default storage.
+                    // TODO: Add a check to prevent files from storage to be deleted.
+                    try
+                    {
+                        File.Delete(p);
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.Error(e, $@"Could not delete file at {p}");
+                    }
+                }
+                catch (Exception e)
+                {
+                    Logger.Error(e, $@"Could not import beatmap set");
+                }
+            
+            // Batch commit with multiple sets to database
+            Import(sets);
         }
 
         public void Import(string path)
         {
+            Import(new [] { path });
+        }
+
+        /// <summary>
+        /// Duplicates content from <paramref name="path"/> to storage and returns a representing <see cref="BeatmapSetInfo"/>.
+        /// Returns null if a representation already exists.

```

Comment:
how about instead of returning `null`, we return the existing BeatmapSetInfo and then check in the actual Import method whether a databased ID is present (and don't insert again)? should simplify this logic.